aten/src/ATen/native/Resize.h

```diff
@@ -46,4 +48,58 @@ inline TensorImpl* resize_impl_cpu_(
   return self;
 }
 
+static inline void checkInBoundsForStorage(
+    IntList size,
+    IntList stride,
+    ptrdiff_t storageOffset,
+    StorageImpl* new_storage) {
+  ptrdiff_t storage_size = 1;
+  for (size_t dim = 0; dim < size.size(); ++dim) {
+    // NB: (a tensor with arbitrary 0 dims)'s storage can have any numel.
+    if (size[dim] == 0) {
+      return;
+    }
+    storage_size += (size[dim] - 1) * stride[dim];
+  }
+  AT_CHECK(
+      storageOffset + storage_size <= new_storage->numel(),
+      "setStorage: sizes ", size, ", strides ", stride, ","
+      " and storage offset ", storageOffset,
+      " requiring a storage size of ", storage_size + storageOffset,
+      " are out of bounds for storage with numel ", new_storage->numel());
+}
+
+/**
+ * Set self's storage to be new_storage with sizes, strides, and storageOffset.
+ * (size, stride, storageOffset) must be in bounds for the new storage.
+ */
+inline void setStorage(
+    TensorImpl* self,
```

Comment:
nit: Consider taking `const Tensor&` here, and getting the storage impl inside this function.